src/server/frontend_wayland/wl_surface.cpp

```diff
@@ -104,23 +104,23 @@ bool mf::WlSurface::synchronized() const
     return role->synchronized();
 }
 
-std::experimental::optional<std::pair<geom::Point, mf::WlSurface*>> mf::WlSurface::transform_point(geom::Point point)
+mf::WlSurface::Position mf::WlSurface::transform_point(geom::Point point)
 {
     point = point - offset_;
     // loop backwards so the first subsurface we find that accepts the input is the topmost one
     for (auto child_it = children.rbegin(); child_it != children.rend(); ++child_it)
     {
         auto result = (*child_it)->transform_point(point);
-        if (result)
+        if (result.is_in_input_region)
             return result;
     }
     geom::Rectangle surface_rect = {geom::Point{}, buffer_size_.value_or(geom::Size{})};
     for (auto& rect : input_shape.value_or(std::vector<geom::Rectangle>{{{}, buffer_size()}}))
```

Comment:
Hrm. Now that I look again, this seems wrong?

From [the protocol](https://gitlab.freedesktop.org/wayland/wayland/blob/master/protocol/wayland.xml#L1479): if the input region is unset, it is treated as infinite.

This is relevant for input events which will be delivered even though they're outside the normal bounds of the surface - eg: pointer drag, touch.

Reply:
What is the problem? If the input region is unset, for any surface/subsurface, the for loop a few lines down loops over one rect the side of the surface:
```
for (auto& rect : input_shape.value_or(std::vector<geom::Rectangle>{{{}, buffer_size()}}))
```
It will then return a `WlSurface::Positoon` with `is_in_input_region` true. Or am I misunderstanding the issue?